### ISSUE TITLE
Why CMS_add1_signer() calls X509_check_private_key()?

### ISSUE THREAD

**Issue by Akretsch** [openssl/openssl#20845]
I wrote a provider wrapping the sign functions (RSA, EC) of an external tool.  So the private key is hidden in an external HSM, but public key and signing certificate are accessible. While executing "openssl cms -provider tw_provider .."  CMS_add1_signer() calls X509_check_private_key() which fails to match the public key and the private key. I did not find a way to "fake" a matching private key in the provider.
Is it really necessary to call X509_check_private_key() before cms signing?
Is an example available how to inject an always matching private key?

**Comment by t8m** [comment]
The provider must support OSSL_FUNC_keymgmt_match_fn call. Not sure if it is a good idea but it could always return true. In that case the X509_check_private_key() should always succeed.

BTW, there was some issue where the match function would not be called properly in previous releases but that was resolved well before the latest release(s) 3.0.8, 3.1.0.

**Comment by Akretsch** [comment]
I have
```
static int tw_keymgmt_match(const void *keydata1, const void *keydata2,
                int selection) {
       return 1;
}
static const OSSL_DISPATCH tw_keymgmt_functions[] = {
DISPATCHALG(OSSL_FUNC_KEYMGMT_FREE, tw_keymgmt_free),
DISPATCHALG(OSSL_FUNC_KEYMGMT_HAS, tw_keymgmt_has),
DISPATCHALG(OSSL_FUNC_KEYMGMT_LOAD, tw_keymgmt_load),
DISPATCHALG(OSSL_FUNC_KEYMGMT_MATCH, tw_keymgmt_match),
DISPATCHALG( 0,NULL) };


```
registered. But its not called. I tried  3.2.0 and 3.0.3.

**Comment by t8m** [comment]
Can you please dump the error stack with ERR_print_errors_fp() to see if there are any errors?

Also the 3.0.3 version is quite old, that might still contain the bug.

**Comment by t8m** [comment]
Hmm... actually the provider must support either export of a key to another provider  or import from other provider to be able to match. In your case I assume only the import OSSL_FUNC_keymgmt_import_fn() might be feasible.

**Comment by t8m** [comment]
805BA515887F0000:error:03000065:digital envelope routines:evp_keymgmt_util_match:different key types:crypto/evp/keymgmt_lib.c:383:

This means the key types differ. What does your provider return as EVP_KEYMGMT_get0_name()?

**Comment by Akretsch** [comment]
Sorry, 
but I searched https://www.openssl.org/docs/man3.0/man7/provider-keymgmt.html for get0_name and similar but didn't find anything. How EVP_KEYMGMT_get0_name should be implemented and registered by my provider?

A little bit confused,
Andreas

**Comment by t8m** [comment]
I am sorry for confusing you. This is the first name from the list of the names the OSSL_ALGORITHM for the keymgmt is registered with.

I.e., for rsa and the default provider it is  PROV_NAMES_RSA (`RSA:rsaEncryption:1.2.840.113549.1.1.1`) as registered at https://github.com/openssl/openssl/blob/master/providers/defltprov.c#L446

**Comment by Akretsch** [comment]
Even if I register my key mangement with `RSA:rsaEncryption:1.2.840.113549.1.1.1`
I get an "crypto/evp/evp_fetch.c:112: OpenSSL internal error: Assertion failed: name_id > 0 && name_id <= METHOD_ID_NAME_MAX". Do you know about a doc describing all the relations and consistency conditions between keytypes, schemes,  OSSL_ALGORITHMs, OSSL_DISPATCHs, decoders and so on?

**Comment by t8m** [comment]
There is no comprehensive provider writers guide unfortunately.

The signature algorithm also needs to be registered with this name (or the query_operation_name function would have to be implemented in the keymgmt).

**Comment by t8m** [comment]
Have you read https://www.openssl.org/docs/man3.0/man7/provider-keymgmt.html ?

**Comment by Akretsch** [comment]
Multiple times :-(

**Comment by Akretsch** [comment]
I registered 
```
static const OSSL_DISPATCH tw_keymgmt_functions[] = {
DISPATCHALG(OSSL_FUNC_KEYMGMT_FREE, tw_keymgmt_free),
DISPATCHALG(OSSL_FUNC_KEYMGMT_HAS, tw_keymgmt_has),
DISPATCHALG(OSSL_FUNC_KEYMGMT_LOAD, tw_keymgmt_load),
DISPATCHALG(OSSL_FUNC_KEYMGMT_MATCH, tw_keymgmt_match),
DISPATCHALG(OSSL_FUNC_KEYMGMT_IMPORT, tw_keymgmt_import),
DISPATCHALG(OSSL_FUNC_KEYMGMT_IMPORT_TYPES, tw_keymgmt_import_types),
		DISPATCHALG(OSSL_FUNC_KEYMGMT_QUERY_OPERATION_NAME,
				tw_keymgmt_query_operation_name),
DISPATCHALG( 0,NULL) };
```
but `tw_keymgmt_query_operation_name` was not called:
```
DEBUG: entering src/tw_provider.c:787:twprovider_query for tw_provider
DEBUG: operation 10 requested from tw_provider
DEBUG: entering src/tw_provider.c:374:tw_keymgmt_load
DEBUG: entering src/tw_provider.c:390:tw_keymgmt_has
DEBUG: entering src/tw_provider.c:596:tw_store_close for tw_provider
DEBUG: entering src/tw_provider.c:445:tw_load_reset for tw_provider
DEBUG: entering src/tw_provider.c:390:tw_keymgmt_has
807B10D5EF7F0000:error:03000065:digital envelope routines:evp_keymgmt_util_match:different key types:crypto/evp/keymgmt_lib.c:383:
807B10D5EF7F0000:error:05800073:x509 certificate routines:ossl_x509_check_private_key:key type mismatch:crypto/x509/x509_cmp.c:416:
807B10D5EF7F0000:error:17000088:CMS routines:CMS_add1_signer:private key does not match certificate:crypto/cms/cms_sd.c:345:

```
`tw_keymgmt_has` returns 1

**Comment by t8m** [comment]
I can only repeat that different key types error means that the keymgmt name, as libcrypto sees it, is different for the private key loaded via your provider and the public key from the default provider.
I do not know what is wrong with your implementation.

**Comment by Akretsch** [comment]
Yes, I agree. The public key comes from the loaded X.509 certificate and has e.g. the type "RSA". I don't have access to my private key, it lives in an HSM. Unfortunately I am not able to fake a (not really existing) "RSA" key in such a way that the libcrypto will use my keymangement for key type and content compare.

**Comment by t8m** [comment]
I do not see why you should not be able to do that. I.e., you should not get this error about different key types, if everything is implemented properly. Your keymgmt must have the RSA in the name if it implements RSA keys, otherwise you can expect various kinds of failures.

**Comment by DDvO** [comment]
Last week I took over the strive to find a solution.
After quite same trial & error, I found a way of getting the CMS signature done by adapting the provider code by @Akretsch.

When using a custom provider/algorithm name, "tw" in this case, for the private key, `X509_check_private_key()`refuses the match because the public key has type "EC". The only thing that seems to works is to use the name "EC" also for loading the certificate (where the public key is included. Adding anything else, such as "EC:tw" or "EC:id-ecPublicKey:1.2.840.10045.2.1" did not work.
I also found that mixing different algorithms (like "EC:RSA") is not accepted at all, so we had to find a different workaround for being able to handle both RSA and EC certs and keys within a single provider.

I also found that for getting CMS signatures through, also the provider signer functions need to refer to the same alg name ("EC" in the above case).

In addition it is needed that the provider key management includes functions for
`OSSL_FUNC_KEYMGMT_IMPORT`, `OSSL_FUNC_KEYMGMT_IMPORT_TYPES`, `OSSL_FUNC_KEYMGMT_GETTABLE_PARAMS`, and `OSSL_FUNC_KEYMGMT_GET_PARAMS`
where the latter provides some `OSSL_PKEY_PARAM_DEFAULT_DIGEST` (such as "SHA256").

**Comment by Akretsch** [comment]
Maybe all the dependencies mentioned by @DDvO should go to an provider implementing guide?

**Comment by DDvO** [comment]
If/when I find the time, I can also provide a PR (or at least suggestions) for extensions and small corrections of the provider documentation in `doc/man7` - well, from my pretty limited experience and user's perspective on providers.

**Comment by Dvergatal** [comment]
Hi guys I'm having the same situation with Vault KMIP PKCS#11 token. But in my case everything is working fine if the certificate passed in `-signer` is self-signed with the token e.g.:
```bash
OPENSSL_CONF=openssl.cnf openssl req -x509 -engine pkcs11 -keyform engine -key "pkcs11:object=development;type=private" -subj "/O=SWUpdate /CN=target" -out development/development.cert.pem
```

If I am creating CSR from the token sign it with my CA than I'm getting this error:
```bash
4017A12B167F0000:error:05800074:x509 certificate routines:X509_check_private_key:key values mismatch:../crypto/x509/x509_cmp.c:408:
4017A12B167F0000:error:17000088:CMS routines:CMS_add1_signer:private key does not match certificate:../crypto/cms/cms_sd.c:316:
```

which is really odd for me... Can anybody may have some suggestions?

**Comment by Dvergatal** [comment]
Ahhh yeah, my bad. It was ineed generated from wrong request... So let's pretend there was no question at all :D

**Comment by DDvO** [comment]
How/where was this (hopefully) solved?

**Comment by t8m** [comment]
This must be solved in the respective provider.

**Comment by t8m** [comment]
If there is something to fix/improve in OpenSSL, please open a new issue describing more precisely the problem and proposed solution.
